### PR TITLE
flow: add a `prometheus.exporter.cadvisor` component

### LINKED ISSUE #3162: Flow component for cadvisor integration
Bring over the static mode cadvisor integration as a Flow component.

This one needs caution because it's currently a singleton integration in static mode, and needs changes to be brought over as not-a-singleton in Flow. See #1861.

### PR DESCRIPTION
**Pull request by tpaschalis**
#### PR Description
This PR adds a `prometheus.exporter.cadvisor` component that runs cAdvisor and allows for scraping container-related metrics.

#### Which issue(s) this PR fixes

Fixes #3162
Updates #1861 

#### Notes to the Reviewer
This whole thing depends on a fork of the google/cadvisor repo over at grafana/cadvisor to allow for non-global instantiation of each 'plugin', to allow for non-singleton components.

The bulk of the work is there; so please take some time to go through the proposed changes on the following PR in case you pick something up we've missed.

https://github.com/grafana/cadvisor/pull/1

#### PR Checklist

- [X] CHANGELOG.md updated
- [x] Documentation added
- [x] Tests updated

### CODE REVIEW

File: docs/sources/flow/reference/components/prometheus.exporter.cadvisor.md [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+Name | Type | Description | Default | Required
+---- | ---- | ----------- | ------- | --------
+`store_container_labels` | `bool` | Whether to convert container labels and environment variables into labels on Prometheus metrics for each container. | `true` | no
+`allowlisted_container_labels` | `list(string)` | Allowlist of container labels to convert to Prometheus labels. | | no
+`env_metadata_allowlist` | `list(string)` | Allowlist of environment variable keys matched with specified prefix that needs to be collected for containers. | | no
+`raw_cgroup_prefix_allowlist` | `list(string)` | List of cgroup path prefix that needs to be collected, even when docker_only is specified. | | no
+`perf_events_config` | `string` | Path to a JSON file containing configuration of perf events to measure. | | no
+`resctrl_interval` | `duration` | Interval to update resctrl mon groups. | `0` | no
+`disabled_metrics` | `list(string)` | List of metrics to be disabled which, if set, overrides the default disabled metrics. | | no
+`enabled_metrics` | `list(string)` | List of metrics to be enabled which, if set, overrides disabled_metrics. | | no
```
Comment (tpaschalis) [inline docs/sources/flow/reference/components/prometheus.exporter.cadvisor.md:35]: AFAICT this list is not about the actual Prometheus metrics being exposed, but rather the `MetricKind` we want to expose (or not). Here's a list of the supported values, I think it's worth documenting them.

https://github.com/grafana/cadvisor/blob/fbd519ba03978d54cb54ea7ed8ab9d6e3dd64590/container/factory.go#L48-L68

File: go.mod [hunk 2]
```diff
@@ -682,6 +682,10 @@ replace (
 
 // TODO(rfratto): remove forks when changes are merged upstream
 replace (
+	// TODO(tpaschalis) this is to remove global instantiation of plugins
+	// and allow non-singleton components.
+	github.com/google/cadvisor => github.com/grafana/cadvisor v0.0.0-20230927082732-0d72868a513e
```
Comment (marctc) [inline go.mod:687]: Do we have a link of the PR already? Would be good to add it here as part of the comment

Comment (tpaschalis) [reply go.mod:687]: Yes, will do. If you want to take a look, the branch is [here](https://github.com/grafana/cadvisor/tree/grafana-v0.47-noglobals)

**Review by erikbaranowski — COMMENTED** [review]
Are you planning to include the wiring in the staticconvert package for this new component?

**Comment by tpaschalis** [comment]
@erikbaranowski whoops, forgot to do that, will have it up on Monday.

**Comment by tpaschalis** [comment]
cc @erikbaranowski I wired in the converter in https://github.com/grafana/agent/pull/5307/commits/9e7d9df1374a1f9341b0989c55fa79db881b3a0f. Can you take a look in case we need anything else, too?

**Comment by erikbaranowski** [comment]
I updated the version of river for an improvement I made and updated the flow argument defaults as we discussed. Check it out and feel free to tune anything.

**Review by erikbaranowski — APPROVED** [review]
LGTM

File: docs/sources/flow/reference/components/prometheus.exporter.cadvisor.md [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+
+## Arguments
+The following arguments can be used to configure the exporter's behavior.
+All arguments are optional. Omitted fields take their default values.
+
+Name | Type | Description | Default | Required
+---- | ---- | ----------- | ------- | --------
+`store_container_labels` | `bool` | Whether to convert container labels and environment variables into labels on Prometheus metrics for each container. | `true` | no
+`allowlisted_container_labels` | `list(string)` | Allowlist of container labels to convert to Prometheus labels. | `[]`  | no
+`env_metadata_allowlist` | `list(string)` | Allowlist of environment variable keys matched with specified prefix that needs to be collected for containers. | `[]` | no
```
Comment (clayton-cornell) [inline docs/sources/flow/reference/components/prometheus.exporter.cadvisor.md:30]:
```suggestion
`env_metadata_allowlist` | `list(string)` | Allowlist of environment variable keys matched with a specified prefix that needs to be collected for containers. | `[]` | no
```

File: docs/sources/flow/reference/components/prometheus.exporter.cadvisor.md [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+`raw_cgroup_prefix_allowlist` | `list(string)` | List of cgroup path prefix that needs to be collected, even when docker_only is specified. | `[]` | no
```
Comment (clayton-cornell) [inline docs/sources/flow/reference/components/prometheus.exporter.cadvisor.md:31]:
```suggestion
`raw_cgroup_prefix_allowlist` | `list(string)` | List of cgroup path prefixes that need to be collected, even when docker_only is specified. | `[]` | no
```

File: docs/sources/flow/reference/components/prometheus.exporter.cadvisor.md [hunk 5]
```diff
@@ -0,0 +1,126 @@
...
+`perf_events_config` | `string` | Path to a JSON file containing configuration of perf events to measure. | `""` | no
```
Comment (clayton-cornell) [inline docs/sources/flow/reference/components/prometheus.exporter.cadvisor.md:32]:
```suggestion
`perf_events_config` | `string` | Path to a JSON file containing the configuration of perf events to measure. | `""` | no
```

**Review by clayton-cornell — APPROVED** [review]
A few minor tweaks to the docs. Otherwise go as-is.

**Pull request merged**